### ISSUE TITLE
Get DND state (as setting DND state) is not working anymore on mac os v.11 (Big Sur)

### ISSUE THREAD

**Issue by RihardKrauz** [sindresorhus/do-not-disturb#9]
getDoNotDisturb is not working anymore on mac os v.11 (Big Sur) since they removed ~/Library/Preferences/ByHost/com.apple.notificationcenterui setting

**Comment by a7madgamal** [comment]
took a look at the issue, here is where I THINK the value is, but I'm stuck trying to update the UI to apply the change (works with logout though)
https://gist.github.com/a7madgamal/9d3518a62c21477a92be461fbd652533

**Comment by PhilipDukhov** [comment]
value for key `dnd_prefs` at `com.apple.ncprefs.plist` is data representations of an other plist! The minus is that it's not updated instantly, takes a couple of seconds. This is how I parse it using swift: 
```
import Foundation

let ncprefsUrl = URL(
    fileURLWithPath: NSString(
        "~/Library/Preferences/com.apple.ncprefs.plist"
    ).expandingTildeInPath
)

func plistFromData(_ data: Data) throws -> [String:Any] {
    try PropertyListSerialization.propertyList(
        from: data,
        format: nil
    ) as! [String:Any]
}

let prefsList = try plistFromData(try Data(contentsOf: ncprefsUrl))
let dndPrefsData = prefsList["dnd_prefs"] as! Data
let dndPrefsList = try plistFromData(dndPrefsData)

// exit with 1 if dnd, 0 otherwise
if let userPref = dndPrefsList["userPref"] as? [String:Any] {
    print(userPref)
    exit(userPref["enabled"] as! Int32)
} else {
    exit(0)
}
```

**Comment by CharlieHess** [comment]
sharing here as many others will surely stumble upon this thread. here's the Objective-C version of parsing that nested plist, also handles the scheduled prefs: https://github.com/felixrieseberg/macos-notification-state/pull/15

**Comment by alse** [comment]
@a7madgamal any luck on updating the UI after the change?

I tried something similar to [commitChanges](https://github.com/sindresorhus/do-not-disturb/blob/master/Sources/do-not-disturb/DoNotDisturb.swift#L10) with a 'com.apple.ncprefs' appid / notification name but couldn't get it to work

**Comment by tiiiecherle** [comment]
> took a look at the issue, here is where I THINK the value is, but I'm stuck trying to update the UI to apply the change (works with logout though)
> https://gist.github.com/a7madgamal/9d3518a62c21477a92be461fbd652533

Thanks for sharing, I found out the commands to trigger the new state. Just use your commands to enable or disable followed by:

```
killall usernoted
killall ControlCenter
```

Works perfectly on macOS 11.1 ;)

Edit: I added a second kill command as it seems necessary if triggering the dnd for the first time after reboot.

**Comment by mellson** [comment]
Nice one @tiiiecherle thanks for posting this!

**Comment by tiiiecherle** [comment]
@mellson You're welcome, I'm glad I could help.

I was trying to find a solution myself when I found this issue and @a7madgamal 's solution.

After quite some time today I also found out how to get the correct values on the command line...

`/usr/libexec/PlistBuddy -x -c 'Print "dnd_prefs"' /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist | xmllint --xpath "string(//data)" - | base64 --decode | xxd -p`

or

`plutil -extract dnd_prefs xml1 -o - /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist | xmllint --xpath "string(//data)" - | base64 --decode | xxd -p`

**Comment by vitorgalvao** [comment]
@tiiiecherle Instead of `/Users/"$USER"`, you can `$HOME`.

**Comment by vitorgalvao** [comment]
For those looking for a standalone solution, I’ve studied options out there (including this thread; thank you all for the previous research) and [made a script](https://github.com/vitorgalvao/tiny-scripts/blob/master/calm-notifications) which works robustly. It doesn’t require additional software (everything ships with macOS) and is only for Big Sur.

If you use Homebrew, you can install it with `brew install vitorgalvao/tiny-scripts/calm-notifications`. Do `calm-notifications --help` to see available options (`on`, `off`, `toggle`, `status`).

**Comment by tiiiecherle** [comment]
@vitorgalvao 
Thanks for the script and the information. I updated the necessary commands for zsh and added them to my [macos system preferences configuration script](https://github.com/tiiiecherle/osx_install_config/blob/b4522acce343ff40a5a09ee5d5ba13a7520fdaa6/11_system_and_app_preferences/11c_macos_preferences_11.sh#L906) for documentation purposes.

For anyone who needs them in a shell script, this works for me in zsh with the new additions:
```
check_dnd_status() {
        DND_STATUS=$(plutil -extract dnd_prefs xml1 -o - /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist | xmllint --xpath "string(//data)" - | base64 --decode | plutil -convert xml1 - -o - | xmllint --xpath 'boolean(//key[text()="userPref"]/following-sibling::dict/key[text()="enabled"])' -)
}
check_dnd_status
    
enable_dnd() {
    defaults read /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist >/dev/null
    DND_HEX_DATA=$(plutil -extract dnd_prefs xml1 -o - /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist | xmllint --xpath "string(//data)" - | base64 --decode | plutil -convert xml1 - -o - | plutil -insert userPref -xml "
    <dict>
        <key>date</key>
        <date>$(date -u +"%Y-%m-%dT%H:%M:%SZ")</date>
        <key>enabled</key>
        <true/>
        <key>reason</key>
        <integer>1</integer>
    </dict> " - -o - | plutil -convert binary1 - -o - | xxd -p | tr -d '\n')
    defaults write com.apple.ncprefs.plist dnd_prefs -data "$DND_HEX_DATA"
    PROCESS_LIST=(
    #cfprefsd
    usernoted
    #NotificationCenter
    )
    while IFS= read -r line || [[ -n "$line" ]] 
	do
	if [[ "$line" == "" ]]; then continue; fi
        i="$line"
        #echo "$i"
        if [[ $(ps aux | grep "$i" | grep -v grep | awk '{print $2;}') != "" ]]
        then
        	killall "$i" && sleep 0.1 && while [[ $(ps aux | grep "$i" | grep -v grep | awk '{print $2;}') == "" ]]; do sleep 0.5; done
	else
		:
	fi
    done <<< "$(printf "%s\n" "${PROCESS_LIST[@]}")"
    #sleep 2
}
enable_dnd

disable_dnd() {
    defaults read /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist >/dev/null
    DND_HEX_DATA=$(plutil -extract dnd_prefs xml1 -o - /Users/"$USER"/Library/Preferences/com.apple.ncprefs.plist | xmllint --xpath "string(//data)" - | base64 --decode | plutil -convert xml1 - -o - | plutil -remove userPref - -o - | plutil -convert binary1 - -o - | xxd -p | tr -d '\n')
    defaults write com.apple.ncprefs.plist dnd_prefs -data "$DND_HEX_DATA"
    PROCESS_LIST=(
    #cfprefsd
    usernoted
    #NotificationCenter
    )
    while IFS= read -r line || [[ -n "$line" ]] 
	do
	if [[ "$line" == "" ]]; then continue; fi
        i="$line"
        #echo "$i"
        if [[ $(ps aux | grep "$i" | grep -v grep | awk '{print $2;}') != "" ]]
        then
        	killall "$i" && sleep 0.1 && while [[ $(ps aux | grep "$i" | grep -v grep | awk '{print $2;}') == "" ]]; do sleep 0.5; done
	else
		:
	fi
    done <<< "$(printf "%s\n" "${PROCESS_LIST[@]}")"
    #sleep 2
}
disable_dnd
```

**EDIT:** I updated the functions after some testing. They seem to work reliably now.

**Comment by karlhorky** [comment]
@tiiiecherle thanks for this! I've used this in this script to automatically stop and start Do Not Disturb when Zoom calls are running, seems to be working well!

https://github.com/karlhorky/do-not-disturb-during-zoom-macos

**Comment by tiiiecherle** [comment]
@karlhorky You are welcome. I'm glad I could help.

Big credits and thanks go to everyone who posted here and gave the needed ideas and input, e.g. @a7madgamal and @vitorgalvao

**Comment by ibc** [comment]
I wonder what's the purpose of wrapping someone else's binary into a EventEmitter is if no maintenance is planned. Thanks, "author".

**Comment by karlhorky** [comment]
Edit, see comments below:

~~I just updated to macOS 12 Monterey and can confirm that [the solution from @tiiiecherle](https://github.com/sindresorhus/do-not-disturb/issues/9#issuecomment-768492417) still works! 🙌~~

**Comment by agross** [comment]
@karlhorky @tiiiecherle For me both enabling and disabling do not work on Monterey. Focus mode/DND stays the way it is set via the UI.

**Comment by karlhorky** [comment]
@agross interesting, wonder what's different between your system and mine.

Here's the script that I am using to trigger that code:

https://github.com/karlhorky/do-not-disturb-during-zoom-macos

**Comment by agross** [comment]
@karlhorky I don't know either, but I have two Macs that show the same behavior. [Getting the DND status](https://github.com/karlhorky/do-not-disturb-during-zoom-macos/blob/master/get-dnd-status.sh) does not change regardless of the Focus Mode setting.

**Comment by nbashkankov** [comment]
@agross - same for me - Monterey has new Focus mode stuff - see #12

**Comment by karlhorky** [comment]
Yeah weirdly enough, [the solution from @tiiiecherle](https://github.com/sindresorhus/do-not-disturb/issues/9#issuecomment-768492417) stopped working for me on Monterey... 😫

Maybe it never worked, and I just tricked myself into thinking it was working...

**Comment by vitorgalvao** [comment]
> Yeah weirdly enough, [the solution from @tiiiecherle](https://github.com/sindresorhus/do-not-disturb/issues/9#issuecomment-768492417) stopped working for me on Monterey...

It’s not weird at all. Programmatically setting Do Not Disturb isn’t simple and Apple keeps breaking methods we come up with. Big Sur required a whole new method, and so will Monterey.

> Maybe it never worked, and I just tricked myself into thinking it was working...

Seems like the likeliest explanation.

**Comment by karlhorky** [comment]
Ah, just came up with an easy and elegant solution using macOS Monterey Shortcuts 🎉

1. Open Shortcuts.app
2. Create a shortcut that ONLY turns on the Focus mode "Do Not Disturb" (rename this to be `dnd-on`)
3. Create a shortcut that ONLY turns off the Focus mode "Do Not Disturb" (rename this to be `dnd-on`)
4. Run `shortcut run dnd-on` on the command line to turn on Do Not Disturb
5. Run `shortcut run dnd-off` on the command line to turn off Do Not Disturb

Screenshots / videos:

![2021-11-18_19-21](https://user-images.githubusercontent.com/1935696/142479316-6713cfc5-3058-4026-8466-c192274d4996.png)

![2021-11-18_19-19](https://user-images.githubusercontent.com/1935696/142479343-a4a39124-3efe-4c94-8e6d-6aacec6eaef6.png)

<img width="367" alt="Screen Shot 2021-11-18 at 19 17 39" src="https://user-images.githubusercontent.com/1935696/142479360-df7e29c9-fe0e-4730-b0bc-f61618e27870.png">


https://user-images.githubusercontent.com/1935696/142481782-06731f84-4012-482a-9cdb-075afa95df59.mp4

**Comment by vitorgalvao** [comment]
I’ve been thinking the same, that Shortcuts would be the solution going forward. At least we no longer need to worry about Apple breaking our solutions every time.

Though that likely does not provide a way to check the *state* of Do Not Disturb.

I’m still on Big Sur so can’t check.

**Comment by karlhorky** [comment]
> Though that likely does not provide a way to check the state of Do Not Disturb.

Right, that makes sense, although apparently [SiriKit can query the state](https://github.com/bdkjones/MuzzleApp/issues/98#issuecomment-953451374) - also [macOS via Mac Catalyst 15.0+ (the `focusStatus` API)](https://developer.apple.com/documentation/sirikit/infocusstatuscenter/3801700-focusstatus?language=objc) (maybe macOS gets this capability too, or already has it but is undocumented)

**Comment by KilAnt** [comment]
> > Though that likely does not provide a way to check the state of Do Not Disturb.
> 
> Right, that makes sense, although apparently [SiriKit can query the state](https://github.com/bdkjones/MuzzleApp/issues/98#issuecomment-953451374) - also [macOS via Mac Catalyst 15.0+ (the `focusStatus` API)](https://developer.apple.com/documentation/sirikit/infocusstatuscenter/3801700-focusstatus?language=objc) (maybe macOS gets this capability too, or already has it but is undocumented)

Does this work for macOS latest version? Anyway, I couldn't find any implementation of Focus Status Center. At least for Objective-C++. I will be glad of any help.

**Comment by vitorgalvao** [comment]
To get Do not Disturb state: https://github.com/vitorgalvao/tiny-scripts/issues/206#issuecomment-974747114

**Comment by KilAnt** [comment]
> To get Do not Disturb state: [vitorgalvao/tiny-scripts#206 (comment)](https://github.com/vitorgalvao/tiny-scripts/issues/206#issuecomment-974747114)

![image](https://user-images.githubusercontent.com/22936434/143868224-4c77e2dc-1984-48fd-b6ed-50e13e116ec3.png)

It works either with a delay, or somehow incorrectly. But after some in it work.

**Comment by tiiiecherle** [comment]
Thanks to everyone for all the updates and the hints.

I also decided to go with the Shortcuts.app when using macOS12 and updated all my [scripts](https://github.com/tiiiecherle/osx_install_config) accordingly.

Checking status and turning on and off seems to work reliably after some testing and a first time configuration of the Shortcuts.app.

**Comment by vitorgalvao** [comment]
[WIth this Shortcut](https://github.com/vitorgalvao/alfred-workflows/blob/master/CalmNotifications/source/Calm%20Notifications.shortcut), you can give it as input `on`, `off`, or a number (representing minutes to stay active). Example:

```bash
shortcuts run 'Calm Notifications' <<< 'on'
```

**Comment by tiiiecherle** [comment]
@vitorgalvao Thanks for the hint. I am not an Alfred Powerpack user, but I will keep it in mind.
The workflow with shortcuts.app and the command line seems to work fine.

**Comment by vitorgalvao** [comment]
> The workflow with shortcuts.app and the command line seems to work fine.

That’s why I shared the link to the relevant file. While I do provide it as part of an Alfred tool, it can be downloaded and used independently.

**Comment by tiiiecherle** [comment]
Thanks, but how and where do I install the workflow? Alfred states on it's website that you have to be a powerpack user to use workflows.

**Comment by arodik** [comment]
Hey folks, I've created a library for shortcut-based focus mode management. 
https://github.com/arodik/macos-focus-mode

@vitorgalvao I've used your shortcut if you don't mind) Thank you 🎉 🎉 🎉

**Comment by john-preston** [comment]
I hope I've fixed that for the next version, need to check in the next 4.6.11 beta when it's out.

**Comment by brookejoseph** [comment]
has anyone been able to find a way of reading which type of focus mode is enabled on the device?

**Comment by sindresorhus** [comment]
Closing this along with the repo. There is nothing left to fix here.

macOS 11 replaced Do Not Disturb with Focus, and Apple ships no public API to read or set it. The private `DoNotDisturb` framework needs an entitlement that third-party apps cannot get, and the private database needs Full Disk Access. Every workaround in this thread, including the `ncprefs` plist method and the `killall usernoted` step, has broken since.

The Shortcuts route works for setting only: make a shortcut with the built-in "Set Focus" action, then `shortcuts run "<shortcut name>"`. Archiving the repo.
